substrate/client/network/src/litep2p/shim/notification/peerset.rs

```diff
@@ -0,0 +1,1276 @@
+			.map(|peer| (*peer, PeerState::Disconnected))
+			.collect::<HashMap<_, _>>();
+
+		// register protocol's commad channel to `Peerstore` so it can issue disconnect commands
+		// if some connected peer gets banned.
+		peerstore_handle.register_protocol(Arc::new(PeersetHandle { tx: cmd_tx.clone() }));
+
+		(
+			Self {
+				protocol,
+				max_out,
+				num_out: 0usize,
+				max_in,
+				num_in: 0usize,
+				reserved_peers,
+				cmd_rx,
+				peerstore_handle,
+				reserved_only,
+				peers,
+				connected_peers,
+				pending_backoffs: FuturesUnordered::new(),
+				next_slot_allocation: Delay::new(SLOT_ALLOCATION_FREQUENCY),
+			},
+			cmd_tx,
+		)
+	}
+
+	/// Report to [`Peerset`] that a substream was opened.
+	///
+	/// Slot for the stream was "preallocated" when the it was initiated (outbound) or accepted
+	/// (inbound) by the local node which is why this function doesn't allocate a slot for the peer.
+	///
+	/// Returns `true` if the substream should be kept open and `false` if the substream had been
+	/// canceled while it was opening and litep2p should close the substream.
+	pub fn report_substream_opened(&mut self, peer: PeerId, direction: traits::Direction) -> bool {
+		log::trace!(
+			target: LOG_TARGET,
+			"{}: substream opened to {peer:?}, direction {direction:?}, reserved peer {}",
+			self.protocol,
+			self.reserved_peers.contains(&peer),
+		);
+
+		let Some(state) = self.peers.get_mut(&peer) else {
+			log::warn!(target: LOG_TARGET, "{}: substream opened for unknown peer {peer:?}", self.protocol);
+			debug_assert!(false);
+			return false
+		};
+
+		match state {
+			PeerState::Opening { direction } => {
+				*state = PeerState::Connected { direction: *direction };
+				self.connected_peers.fetch_add(1usize, Ordering::Relaxed);
+				true
+			},
+			// litep2p doesn't support the ability to cancel an opening substream so if the
+			// substream was closed while it was opening, it was marked as canceled and if the
+			// substream opens succesfully, it will be closed
+			PeerState::Canceled { direction } => {
+				log::trace!(
+					target: LOG_TARGET,
+					"{}: substream to {peer:?} is canceled, issue disconnection request",
+					self.protocol,
+				);
+
+				self.connected_peers.fetch_add(1usize, Ordering::Relaxed);
+				*state = PeerState::Closing { direction: *direction };
+				false
+			},
+			state => {
+				panic!("{}: invalid state for open substream {peer:?} {state:?}", self.protocol);
+			},
+		}
+	}
+
+	/// Report to [`Peerset`] that a substream was closed.
+	///
+	/// If the peer was not a reserved peer, the inbound/outbound slot count is adjusted to account
+	/// for the disconnected peer. After the connection is closed, the peer is chilled for a
+	/// duration of [`DEFAULT_BACKOFF`] which prevens [`Peerset`] from establishing/accepting new
+	/// connections for that time period.
+	///
+	/// Reserved peers cannot be disconnected using this method and they can be disconnected only if
+	/// they're banned.
```

Comment:
Could you elaborate on this comment? As I see it, `report_substream_closed` is a result of `litep2p` notifying us that a substream was closed. Why can't it happen for a reserved node?

Reply:
I think it's just a very old comment back when that function did something other than it does now. I'll remove it